crates/cli/benches/benchmark.rs

```diff
@@ -0,0 +1,119 @@
+use criterion::{criterion_group, criterion_main, Criterion};
+use num_format::{Locale, ToFormattedString};
+use std::{error::Error, fs, path::Path, process::Command};
+use wasi_common::pipe::{ReadPipe, WritePipe};
+use wasmtime::{Config, Engine, Linker, Module, Store};
+use wasmtime_wasi::sync::WasiCtxBuilder;
+
+struct Function {
+    name: String,
+    wasm_bytes: Vec<u8>,
+    payload: Vec<u8>,
+    engine: Engine,
+}
+
+impl Function {
+    pub fn new(function_dir: &Path) -> Result<Function, Box<dyn Error>> {
+        let name = function_dir
+            .file_name()
+            .ok_or("Path terminates in ..")?
+            .to_str()
+            .ok_or("Function file name contains invalid unicode")?
+            .to_string();
+
+        let wasm_path = function_dir.join("index.wasm");
+        execute_javy(&function_dir.join("index.js"), &wasm_path);
+
+        Ok(Function {
+            name,
+            wasm_bytes: fs::read(wasm_path)?,
+            payload: fs::read(function_dir.join("input.json"))?,
+            engine: Engine::new(&Config::default())?,
+        })
+    }
+
+    pub fn compile(&self) -> Result<Vec<u8>, Box<dyn Error>> {
+        let module = Module::new(&self.engine, &self.wasm_bytes)?.serialize()?;
+        Ok(module)
+    }
+
+    pub fn run_precompiled(&self, elf_js_module: &[u8]) -> Result<(), Box<dyn Error>> {
+        let js_module = unsafe { Module::deserialize(&self.engine, elf_js_module) }?;
+        self.run(&js_module)?;
+        Ok(())
+    }
+
+    pub fn run_uncompiled(&self) -> Result<(), Box<dyn Error>> {
+        let js_module = Module::new(&self.engine, &self.wasm_bytes)?;
+        self.run(&js_module)?;
+        Ok(())
+    }
+
+    fn run(&self, js_module: &Module) -> Result<(), Box<dyn Error>> {
+        let mut linker = Linker::new(&self.engine);
+        let stdout = WritePipe::new_in_memory();
+        let wasi = WasiCtxBuilder::new()
+            .stdin(Box::new(ReadPipe::from(&self.payload[..])))
+            .stdout(Box::new(stdout.clone()))
+            .stderr(Box::new(stdout))
+            .build();
+        wasmtime_wasi::add_to_linker(&mut linker, |s| s)?;
```

Comment:
According to the [spec](https://webassembly.github.io/spec/core/intro/overview.html)  execution includes _instantiation_ and _invocation_ and instantiation defined -- in very simple terms -- as name resolution for imports, which should be already available at the time of instantiating, which is what the linker is doing here. That's mainly the reason I generally don't take this phase into account. Also in practice, if you want to have multiple instances per Store, you only need to setup this once.

Reply:
So I moved the linker and store creation into a setup method that is called by a Criterion method since the input (that is, `Function`) is passed to the bench closure as immutably borrowed and the linker and store need to be mutated inside `run` as part of module instantiation.

The `setup` method on `Function` is immutable so I think invoking the `setup` method as part of the setup closure should be okay.